BERT pretraining

piotrekp1 opened the pull request:
Hi! this commit adds:
NSP preprocessing
Masking procedure,
MLM head
and pretraining head that combines MLM with NSP

NSP is a little hacky, that's because I wanted to use function from t5 library, but it operates on tfds dataset and returns concatenated string with 2 sentences for each row. NSPInputs handles the interface differences.

henrykmichalewski commented on trax/data/tf_inputs.py:
```diff
@@ -1107,6 +1112,42 @@ def bert_double_sentence_preprocess(batch):
     yield value_vector, segment_embs, segment_embs, label, np.int64(1)
 
 
+def mask_random_tokens(batch, vocab_size=30522, masking_prob=0.15, merge_with_loss=True):
```
Can you extend the comment and link the mask_random_tokens method in the BERT implementation you have in mind?

henrykmichalewski commented on trax/models/research/bert.py:
```diff
@@ -71,8 +71,39 @@ def BERTRegressionHead():
+  ])
+
+@gin.configurable()
+def BERTPretrainingLoss():
```
Can you split the losses, that is make each of them optional?

piotrekp1 replied:
Losses are already optional (You can use just MLMhead or just ClassifierHead with WeightedCrossEntropy loss and it will work). However I did not implement choosing random sentence from corpus document for MLM training without adding NSP loss. I'll add it and add gin file for training only MLM as an example

henrykmichalewski reviewed (changes requested):
Thanks for the PR!

afrozenator commented on trax/data/__init__.py:
```diff
@@ -44,7 +44,9 @@ def data_configure(*args, **kwargs):
 Shuffle = data_configure(inputs.Shuffle)
 shuffle = inputs.shuffle
 TFDS = data_configure(tf_inputs.TFDS)
+NSPInputs = data_configure(tf_inputs.NSPInputs)
```
Let's spell it out fully as `NextSentencePredictionInputs`?

afrozenator commented on trax/data/tf_inputs.py:
```diff
@@ -1107,6 +1112,42 @@ def bert_double_sentence_preprocess(batch):
+  `masking_prob` random tokens are selected out of which:
+    80% are replaced with [MASK] token
+    10% are replaced with random token
+    10% are leaved with original token unchanged
```
nit: s/leaved/left

afrozenator commented on trax/data/tf_inputs.py:
```diff
@@ -1107,6 +1112,42 @@ def bert_double_sentence_preprocess(batch):
+  ids of the tokens are from original lowercase vocab
+  """
+
+  for token_ids, *row_rest in batch:
```
Could you comment either in the docstring or comments what `batch` is, like type, shape etc

afrozenator commented on trax/data/tf_inputs.py:
```diff
@@ -1107,6 +1112,42 @@ def bert_double_sentence_preprocess(batch):
+  for token_ids, *row_rest in batch:
+    y = token_ids.copy()
+
+    is_special_token = np.logical_or(token_ids == 101, token_ids == 102) # CLS and SEP tokens
```
Let's make the CLS and SEP token-ids arguments in this function -- these maybe different token ids for different vocabularies, so they should be settable in the argument to this function.

piotrekp1 replied:
Thank you! That's a very simple and clean solution. I knew these magic numbers were a problem, but somehow I thought only about reading these special ids from vocabulary which was problematic in some cases, so I decided to leave them to deal with them some other day. Your solution is obviously much better.

piotrekp1 replied:
I added all special ids as arguments for function

afrozenator commented on trax/data/tf_inputs.py:
```diff
@@ -1107,6 +1112,42 @@ def bert_double_sentence_preprocess(batch):
+    mask_threshold, random_token_threshold = 0.8 * masking_prob, 0.9 * masking_prob
+    # change 80 % of tokens to [MASK]
+    mask_tokens = prob_scores < mask_threshold
+    token_ids[mask_tokens] = 103
```
lets also make this MASK id i.e. 103 an argument

afrozenator commented on trax/data/tf_inputs.py:
```diff
@@ -1107,6 +1112,42 @@ def bert_double_sentence_preprocess(batch):
+
+    # change 10% of tokens to random token
+    random_tokens = np.logical_and(prob_scores > mask_threshold, prob_scores < random_token_threshold)
+    token_ids[random_tokens] = np.random.randint(999, vocab_size, random_tokens.sum())
```
Can you explain a little bit in comments what's going on here, why is 999 special?

piotrekp1 replied:
999 is the first token id in standard bert_base_lowercase vocab. I will add this to arguments just like other special ids.

Here for 10% of tokens selected for prediction are changed to random tokens.
This is done in this way because of the sampling methodology i used. Instead of using 2 random samples: 
1) First whether given will be used to evaluation (0.15 probability in standard version)
2) split selected 15% of tokens into 3 groups: (0.8 MASK, 0.1 random token, 0.1 leave original)
Which is the way it is described.
I do one split which splits it into 4 groups: (p_choose is 0.15 in original version)
  (0.8 * p_choose) MASK, 
  (0.1 * p_choose) random token, 
  (0.1 * p_choose) original tokens, 
  (1 - p_choose) tokens not predicted.
Which uses just one sampling and is the same experiment. 
I'll add some comment to the docstring about it for clarity

piotrekp1 replied:
I checked out the original bert code and realized that in original bert 15% of tokens are always chosen (instead of choosing for each token independently with 15% probability whether it is involved in masking). So I rewritten the code for consistency with original method. I also extended the documentation quite significantly

afrozenator reviewed (changes requested):
Hi @piotrekp1 - Thanks for sending the PR! I have a few comments, please take a look and let me know!

I'm not sure if all my comments will show up ....

piotrekp1 commented:
Thank you for all the comments. I will resolve them and let you know.

afrozenator commented:
Thanks @piotrekp1 - let me know when its ready for another round of review! Thanks for adding this!

piotrekp1 commented:
I added new commit with all suggested changes. I also changed CreateBertInputs function a little, because I realized it wasn't working for inputs without labels. It would be great if you could check whether now there's anything that should be added or changed.

afrozenator commented on trax/data/tf_inputs.py:
```diff
@@ -240,14 +240,19 @@ def TFDS(  # pylint: disable=invalid-name
-  """Returns an iterator of numpy arrays representing the dataset."""
+  """Returns an iterator of numpy arrays representing the dataset.
+  Args:
+     tfds_preprocess_fn: function that transforms tensorflow dataset applied before changing dataset to generator
```
Could we either add the other args, or delete this arg and just leave the docstring?

piotrekp1 replied:
I added the other args in the docstring. It might be not clear what 'tfds_preprocess_fn' is just from the name, so I think it's better to include all args than none.

afrozenator commented on trax/data/tf_inputs.py:
```diff
@@ -1090,30 +1095,136 @@ def download_and_prepare(dataset_name, data_dir):
+    for sent1, label in batch:
+      value_vector = np.concatenate(([CLS_id], sent1, [SEP_id]))
+      segment_embs = np.zeros(sent1.shape[0] + 2, dtype=np.int32)
+      yield value_vector, segment_embs, segment_embs, label, np.int64(1)
```
why `np.int64(1)` and not `np.int32(1)` -- will it work with int32? If so, lets do that.

piotrekp1 replied:
Valid point, I changed it to np.uint32 (these weights are never negative anyway). I also changed segment_embs to np.uint16, because they are tiny (right now never even greater than 1)

afrozenator replied:
Let's not do uints -- I'm not sure if we handle that correctly across tf/jax/tf-np etc ... Or what will happen if something downstream assumes a uint32 is an int32 ... let's stick to np.int32 everywhere we can. Thanks!

piotrekp1 replied:
okay, i changed them all to int32 :-)

afrozenator commented on trax/data/tf_inputs.py:
```diff
@@ -1090,30 +1095,136 @@ def download_and_prepare(dataset_name, data_dir):
+      segment_embs = np.zeros(sent1.shape[0] + 2, dtype=np.int32)
+      yield value_vector, segment_embs, segment_embs, label, np.int64(1)
+  else:
+    for sent1, *_ in batch: # row is a tuple with 1 element
```
@piotrekp1 - thanks for the comment (and teaching me something new) but this syntax is a quite confusing to me, can we be explicit here?

Something like:
```
for row in batch:
  # row is a tuple with 1 element
  sent1 = row[0]
```

Or perhaps -- does this even work?
```
for (sent1,) in batch: # row is a tuple with 1 element
```

piotrekp1 replied:
Actually (sent1, ) worked! Thanks, now you too taught me something about this syntax :)

afrozenator commented on trax/data/tf_inputs.py:
```diff
@@ -1090,30 +1095,136 @@ def download_and_prepare(dataset_name, data_dir):
-  else:
-    return bert_single_sentence_preprocess
+def CreateBertInputs(double_sentence=True, labeled=True, CLS_id=101, SEP_id=101):
+  foo = BertDoubleSentenceInputs if double_sentence else BertSingleSentenceInputs
```
s/foo/bert_fn or something else but not foo :-)

piotrekp1 replied:
I changed it to "function". Is that better? :) Just kidding, changed it to bert_inputs_fn

afrozenator commented on trax/data/tf_inputs.py:
```diff
@@ -1090,30 +1095,136 @@ def download_and_prepare(dataset_name, data_dir):
+      - replaced with random token with 10% probability
+      - unchanged with 10%
+  Follows standard procedure of masking in BERT found in
+  create_masked_lm_predictions-https://github.com/google-research/bert/blob/master/create_pretraining_data.py, line 342.
```
We haven't switched on the linter externally due to some issues, but can you keep the comments (and code lines) to 80 characters?

Thanks!

piotrekp1 replied:
Oh, sorry! I have a vertical line in my IDE that shows me the limit of 80 chars. It looked it's still less, I guess even coming that close to 80 chars was unnecessary.

afrozenator commented on trax/data/tf_inputs.py:
```diff
@@ -1090,30 +1095,136 @@ def download_and_prepare(dataset_name, data_dir):
+    token_ids[random_token_ids] = np.random.randint(vocab_start_id, vocab_size, random_token_ids.shape[0])
+
+    # leave 10% unchanged
+
```
delete the extra line?

afrozenator commented on trax/supervised/configs/bert_pretraining.gin:
```diff
@@ -0,0 +1,95 @@
+# Copyright 2020 The Trax Authors.
```
shouldn't this file be named bert_nsp_pretraining.gin?

piotrekp1 replied:
Actually no, because it uses both NSP and MLM as in original bert, but for clarity i changed the nomenclature a bit. Now i have bert_pretraining.gin (both losses), bert_pretraining_onlymlm.gin and bert_pretraining_onlynsp.gin and I think it should be clear.

afrozenator commented on trax/supervised/configs/bert_mlm_pretraining.gin:
```diff
@@ -0,0 +1,95 @@
+# Copyright 2020 The Trax Authors.
```
together with the other gin file -- these have a lot in common -- can we rename the below gin to bert_nsp_pretraining.gin and extract the common part to bert_pretraning.gin

and include bert_pretraining.gin in both these files?

For an example see https://github.com/google/trax/blob/master/trax/supervised/configs/sparse_c4_pretrain_16gb_adafactor.gin which includes c4.gin - https://github.com/google/trax/blob/master/trax/supervised/configs/c4.gin

piotrekp1 replied:
Thanks! I wasn't aware you can do modularity in gins. I took it step further and extracted bert.gin and refactored all bert-related gins so that they include bert.gin.

afrozenator reviewed (changes requested):
Hi @piotrekp1 - this is much better, just a few more changes and this should be looking quite good.

However as a major comment -- can I request you to write a few tests? They should be easy if you see the existing tests in tf_inputs_test.py or other test files in the same directory -- otherwise there is a high chance things will break.

Thanks!

piotrekp1 commented:
Thank you @afrozenator very much for the review, it's very instructive. I added tests for masking and NextSentencePredictionInputs functions as well as all other suggestions.

afrozenator commented on trax/supervised/configs/bert.gin:
```diff
@@ -0,0 +1,70 @@
+import trax.supervised.trainer_lib
+import trax.layers.metrics
+
+# See https://www.tensorflow.org/datasets/catalog/glue -- valid dataset_name
```
does this comment block make sense for bert.gin? given that dataset_name doesn't exist here ... its OK to replicate across the child gin files I think.

afrozenator commented on trax/supervised/configs/bert_glue_regression.gin:
```diff
@@ -29,24 +23,18 @@ import trax.layers.metrics
 
 dataset_name = 'glue/qnli'
 
+include 'bert.gin'
```
same comment as in the previous gin file here.

afrozenator commented on trax/supervised/configs/bert_glue_classification.gin:
```diff
@@ -12,41 +12,29 @@
-  @data.Shuffle(),
-  @data.PadToLength(),
-  @data.TruncateToLength(),
+  %bert_stream_core,
```
duplicating `bert_stream_core` inline (i.e. as it was before) was fine ... because now mostly one will have to look back and forth between files to immediately see what the pipeline is ... but its OK for now.

Even commenting here that "bert_stream_core - tokenizes, creates berk inputs, shuffles, pads and truncates the sequence" will be good here.

piotrekp1 replied:
I see. I removed the bert_stream_core altogether and moved it inline as before. I thought that commenting with what it does might lead to inconsistencies if bert_stream_core is changed and comment is not

afrozenator commented on trax/supervised/configs/bert_pretraining.gin:
```diff
@@ -0,0 +1,56 @@
+
+dataset_name = 'wiki40b'
+
+include 'bert.gin'
```
same as before move bert.gin to before/after the include block

afrozenator commented on trax/supervised/configs/bert_pretraining_onlymlm.gin:
```diff
@@ -0,0 +1,54 @@
+
+dataset_name = 'wiki40b'
+
+include 'bert.gin'
```
same here

afrozenator commented on trax/supervised/configs/bert_pretraining_onlynsp.gin:
```diff
@@ -0,0 +1,50 @@
+
+dataset_name = 'wiki40b'
+
+include 'bert.gin'
```
same here

afrozenator commented on trax/data/tf_inputs_test.py:
```diff
@@ -591,21 +591,50 @@ def test_create_bert_inputs(self):
       self.assertEqual(label, np.int64(1))
       self.assertEqual(weight, np.int64(1))
 
+  def test_mask_random_tokens(self):
```
a few things here:

- it again seems like 103 is MASK?
- why is 101 special (and why it isn't being masked), let's give it its name "CLS"? what?
- we are checking that sum(out == 103) <= 15 ... but also make sure that it is > 0?

piotrekp1 replied:
i gave all magic numbers names. CLS isn't masked because in real scenario it marks the beginning of the sentence and is used to get sentence embedding. Here there are a lot of CLS to test whether function won't mask CLS and whether the number of masked tokens won't be infected by that.

sum(out == MASK_id) can be equal to zero. For each candidated it is randomly chosen whether it is masked (80% probability) changed to random token (10%) and unchanged (10%). I didn't set the seed here so it might (although unlikely) happen that none of the tokens is are masked.

afrozenator commented on trax/data/tf_inputs_test.py:
```diff
@@ -591,21 +591,50 @@ def test_create_bert_inputs(self):
+    # no more than 15 tokens with mask
+    self.assertLessEqual((out == 103).sum(), 15)
+
+  def test_BertNextSentencePredictionInputs(self):
```
let's name as `test_bert_next_sen .... inputs()`

Also how does the test work? I don't see wiki40b's input files in the testdata ...

afrozenator reviewed (commented):
Thanks @piotrekp1 !! We're almost there!

piotrekp1 commented:
Thank you for your patience @afrozenator. I changed all the things you requested, could you take a look?
